zerocopy-derive/src/lib.rs

```diff
@@ -405,40 +409,53 @@ fn derive_unaligned_inner(ast: &DeriveInput) -> proc_macro2::TokenStream {
 // A struct is `TryFromBytes` if:
 // - all fields are `TryFromBytes`
 
-fn derive_try_from_bytes_struct(ast: &DeriveInput, strct: &DataStruct) -> proc_macro2::TokenStream {
-    let extras = Some({
-        let fields = strct.fields();
-        let field_names = fields.iter().map(|(name, _ty)| name);
-        let field_tys = fields.iter().map(|(_name, ty)| ty);
-        quote!(
-            // SAFETY: We use `is_bit_valid` to validate that each field is
-            // bit-valid, and only return `true` if all of them are. The bit
-            // validity of a struct is just the composition of the bit
-            // validities of its fields, so this is a sound implementation of
-            // `is_bit_valid`.
-            fn is_bit_valid<A: ::zerocopy::pointer::invariant::Aliasing + ::zerocopy::pointer::invariant::AtLeast<::zerocopy::pointer::invariant::Shared>>(
-                mut candidate: ::zerocopy::Maybe<Self, A>
-            ) -> bool {
-                true #(&& {
-                    // SAFETY:
-                    // - `project` is a field projection, and so it addresses a
-                    //   subset of the bytes addressed by `slf`
-                    // - ..., and so it preserves provenance
-                    // - ..., and `*slf` is a struct, so `UnsafeCell`s exist at
-                    //   the same byte ranges in the returned pointer's referent
-                    //   as they do in `*slf`
-                    let field_candidate = unsafe {
-                        let project = |slf: *mut Self|
-                            ::zerocopy::macro_util::core_reexport::ptr::addr_of_mut!((*slf).#field_names);
-
-                        candidate.reborrow().project(project)
-                    };
-
-                    <#field_tys as ::zerocopy::TryFromBytes>::is_bit_valid(field_candidate)
-                })*
-            }
-        )
-    });
+fn derive_try_from_bytes_struct(
+    ast: &DeriveInput,
+    strct: &DataStruct,
+    top_level: Trait,
+) -> proc_macro2::TokenStream {
+    let extras = if top_level == Trait::FromBytes {
```

Comment:
Oops, this check is not sufficient! In particular, `#[derive(FromBytes)]` applied to a type with generics will compile successfully, but that doesn't mean that the type is always `FromBytes` - the generated impl is bounded by its generic field types also being `FromBytes`.

It *might* be sufficient to do `top_level == Trait::FromBytes` *and* check that the type has no generics. However, this would rely for soundness on compilation always failing for non-`FromBytes` types. E.g., what if non-`FromBytes` types just result in the `FromBytes` impl not being available, but the compilation still succeeding?

What we probably should do is to add an explicit `Self: FromBytes` check in the body of the generated `is_bit_valid`. I'm pretty sure that's foolproof.

(Same applies to derives for unions and enums)

Reply:
> It _might_ be sufficient to do `top_level == Trait::FromBytes` _and_ check that the type has no generics. However, this would rely for soundness on compilation always failing for non-`FromBytes` types. E.g., what if non-`FromBytes` types just result in the `FromBytes` impl not being available, but the compilation still succeeding?
> 
> What we probably should do is to add an explicit `Self: FromBytes` check in the body of the generated `is_bit_valid`. I'm pretty sure that's foolproof.

Done - the latest version does both of these.